[Math] Align frames to pixel boundaries

romoore opened the pull request:
Adding a new function to realign a CGRect (frame) to pixel boundaries
based on the screen scale. After a review of the catalog, I'm updating
the areas where I found misaligned frames.

|Component|Before|After|
|------------|-------|-----|
|Slider | ![slider-blur](https://user-images.githubusercontent.com/1753199/28936350-75aa9bba-7855-11e7-82af-4cecd0f5fbfc.png)|![slider-pixel](https://user-images.githubusercontent.com/1753199/28936358-791350c6-7855-11e7-84eb-513fcbd92fab.png)
|Feature Highlight|![feature-blur](https://user-images.githubusercontent.com/1753199/28936373-8c1b592a-7855-11e7-9895-dddbe5e5faf9.png)|![feature-pixel](https://user-images.githubusercontent.com/1753199/28936380-90aa3146-7855-11e7-891b-39388284dafe.png)
|Navigation Bar (Buttons)|![navbar_blur](https://user-images.githubusercontent.com/1753199/28936395-9e4212d8-7855-11e7-8166-e6721300c4f1.png)|![navbar_pixel](https://user-images.githubusercontent.com/1753199/28936406-a830ee2c-7855-11e7-9285-a948d1045b35.png)





Closes #1723

Commit:
Align frames to pixel boundaries

Adding a new function to realign a CGRect (frame) to pixel boundaries
based on the screen scale. After a review of the catalog, I'm updating
the areas where I found misaligned frames.

Closes #1723

ajsecord commented on components/Slider/examples/SliderCompareExampleViewController.m:
```diff
@@ -17,6 +17,7 @@
 #import <UIKit/UIKit.h>
 
 #import "MaterialSlider.h"
+#import "MaterialMath.h"
```
Sort imports...

romoore replied:
Done

ajsecord commented on components/private/Math/src/MDCMath.h:
```diff
@@ -94,3 +94,14 @@ static inline CGFloat MDCSqrt(CGFloat value) {
 #endif
 }
+
+static inline CGRect MDCRectIntegral(CGRect rect) {
```
This isn't as obvious as just `MDCSqrt` or whatever, please document.

Commit:
Sorting imports, documenting new function

ajsecord commented on components/private/Math/src/MDCMath.h:
```diff
@@ -94,3 +94,20 @@ static inline CGFloat MDCSqrt(CGFloat value) {
+
+/**
+ Expand `rect' to the smallest rect containing it with integral origin and
+ size.
```
Not really true. Can you please document the `scale` functionality? I'm assuming that this is meant to round to pixel boundaries?

romoore replied:
Done.

Commit:
Tweaking wording of RectIntegral method

romoore commented:
While adding unit tests, I found the current approach does not match CGRectIntegral.

Commit:
Adding unit test, fixing implementation to match CGRectIntegral behavior

willlarche commented on components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.m:
```diff
@@ -130,7 +131,8 @@ - (void)layoutButtonsInRange:(NSRange)range around:(CGFloat)centerX {
     button.frame = CGRectOffset(button.frame, 0, verticalCenterY);
-    label.frame = CGRectOffset(label.frame, 0, verticalCenterY);
+    label.frame = MDCRectAlignToScale(CGRectOffset(label.frame, 0, verticalCenterY),
+                                      [[UIScreen mainScreen] scale]);
```
scale is a property. Best to access with .

romoore replied:
Done

willlarche commented on components/Buttons/src/MDCButton.m:
```diff
@@ -299,6 +299,7 @@ - (void)layoutSubviews {
   } else {
     _inkView.frame = self.bounds;
   }
+  self.titleLabel.frame = MDCRectAlignToScale(self.titleLabel.frame, [[UIScreen mainScreen] scale]);
```
Same about scale, tho I know it's inconsistent in the codebase. We can correct them as we go along or feel free to do a PR with a batch change. Same for bounds.

romoore replied:
Done

willlarche commented on components/FeatureHighlight/examples/supplemental/FeatureHighlightExampleSupplemental.m:
```diff
@@ -74,8 +75,8 @@ - (void)viewDidLayoutSubviews {
-  self.infoLabel.frame = CGRectMake(self.view.frame.size.width / 2 - labelSize.width / 2, 20,
-                                    labelSize.width, labelSize.height);
+  self.infoLabel.frame = MDCRectAlignToScale(
+                CGRectMake(self.view.frame.size.width / 2 - labelSize.width / 2, 20, labelSize.width, labelSize.height), [[UIScreen mainScreen] scale]);
```
Verify that this is the correct spacing with format_all

romoore replied:
Done

willlarche commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -294,7 +294,7 @@ - (void)layoutSubviews {
   }
 
-  CGRect titleFrame = (CGRect){titlePos, titleSize};
+  CGRect titleFrame = MDCRectAlignToScale((CGRect){titlePos, titleSize}, [[UIScreen mainScreen] scale]);
```
.scale

romoore replied:
Done

willlarche commented on components/private/ThumbTrack/src/MDCThumbTrack.m:
```diff
@@ -620,6 +620,9 @@ - (void)updateViewsMainIsAnimated:(BOOL)animated
     if ([_delegate respondsToSelector:@selector(thumbTrack:stringForValue:)]) {
       _valueLabel.text = [_delegate thumbTrack:self stringForValue:_value];
+      if (CGRectGetWidth(_valueLabel.frame) > 1) {
+        _valueLabel.frame = MDCRectAlignToScale(_valueLabel.frame, [[UIScreen mainScreen] scale]);
```
.scale

romoore replied:
Done

willlarche commented on components/private/Math/src/MDCMath.h:
```diff
@@ -94,3 +94,21 @@ static inline CGFloat MDCSqrt(CGFloat value) {
+  CGPoint newOrigin = CGPointMake(MDCFloor(rect.origin.x * scale) / scale,
+                                  MDCFloor(rect.origin.y * scale) / scale);
+  CGSize adjustWidthHeight = CGSizeMake(rect.origin.x - newOrigin.x,
+                                         rect.origin.y - newOrigin.y);
```
alignment

romoore replied:
Done

Commit:
Changing to .scale (property) and reformatting code

ianegordon commented on components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.m:
```diff
@@ -130,7 +131,8 @@ - (void)layoutButtonsInRange:(NSRange)range around:(CGFloat)centerX {
     button.frame = CGRectOffset(button.frame, 0, verticalCenterY);
-    label.frame = CGRectOffset(label.frame, 0, verticalCenterY);
+    label.frame = MDCRectAlignToScale(CGRectOffset(label.frame, 0, verticalCenterY),
+                                      [UIScreen mainScreen].scale);
```
Note: [UIScreen mainScreen] is now a class property and could be accessed by UIScreen.mainScreen.
We should probably discuss whether we want to adopt class properties at our next weekly.

romoore replied:
We may also want to consider avoiding `mainScreen` altogether and using [UIView.window.screen](https://developer.apple.com/documentation/uikit/uiwindow/1621597-screen) instead.

ianegordon commented on components/FeatureHighlight/src/private/MDCFeatureHighlightView+Private.m:
```diff
@@ -285,7 +286,7 @@ - (void)layoutSubviews {
   CGFloat leftTextBound = kMDCFeatureHighlightTextPadding;
   CGFloat rightTextBound = self.frame.size.width - MAX(titleSize.width, detailSize.width) -
-      kMDCFeatureHighlightTextPadding;
+                           kMDCFeatureHighlightTextPadding;
```
nit: alignment.  I would expect a 4 character indent.

romoore replied:
Done

ianegordon commented on components/private/Math/src/MDCMath.h:
```diff
@@ -94,3 +94,20 @@ static inline CGFloat MDCSqrt(CGFloat value) {
+    return CGRectIntegral(rect);
+  }
+  CGPoint newOrigin =
+      CGPointMake(MDCFloor(rect.origin.x * scale) / scale, MDCFloor(rect.origin.y * scale) / scale);
```
This currently only works with standardized rects.
We likely want
CGFloat xOffset = MDCFloor(CGRectGetXOffset(rect) * scale) / scale;
CGFloat yOffset = MDCFloor(CGRectGetYOffset(rect) * scale) / scale;

There is some additional logic around widths.

Prefer CGRect accessor macros over rect.origin or rect.size unless you have explicitly called CGRectStandardize()

romoore replied:
Done

ianegordon commented on components/private/Math/tests/unit/MDCMathTests.m:
```diff
@@ -0,0 +1,52 @@
+  CGRect misalignedRect = CGRectMake(0.45, 0.78, 1.01, 5.98);
+  CGRect alignedScale1Rect = CGRectMake(0, 0, 2, 7);
+  CGRect alignedScale2Rect = CGRectMake(0, 0.5, 1.5, 6.5);
+  CGRect alignedScale3Rect = CGRectMake(1.0 / 3.0, 2.0 / 3.0, 4.0 / 3.0, 19.0 / 3.0);
```
Consider adding a test on a non-standardized CGRect.

CGRect nonstandardRect = CGRectMake (100.0, 100.0, -40, -20)

Ensure we are still fully enclosing the original rect, even if the height and/or width is negative.

romoore replied:
Done

Commit:
Adding test for non-standard rect, formatting

Additional test to verify that a non-standardized rect (negative
height/width) will still be correctly aligned.  Enhanced unit tests to
ensure that the resulting rects will fully contain the original rects.

Commit:
Fixing indentation

ianegordon commented:
LGTM.  Please merge once Travis gives the OK.
